File: mylibrary/src/main/java/com/xiaolu/mylibrary/net/RxScheduler.java

package com.xiaolu.mylibrary.net;


import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;


/**
 * description：RXjava 线程调度
 */

public class RxScheduler {
    /**
     * 统一线程处理
     *
     * @param <T> 指定的泛型类型
     * @return FlowableTransformer
     */

    public static <T> ObservableTransformer<T, T> floIoMain(int timeOut, TimeUnit unit) {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .debounce(timeOut, unit);
    }

    public static <T> ObservableTransformer<T, T> floIoIo(int timeOut, TimeUnit unit) {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .debounce(timeOut, unit);
    }


    /**
     * 统一线程处理
     *
     * @param <T> 指定的泛型类型
     * @return ObservableTransformer
     */

    public static <T> ObservableTransformer<T, T> obsIoMain(int timeOut, TimeUnit unit) {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .debounce(timeOut, unit);
    }

}

File: mylibraryKotlin/src/main/java/com/xiaolu/mylibrarykotlin/net/RxScheduler.kt
package com.xiaolu.mylibrarykotlin.net

import io.reactivex.ObservableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit

/**
 * description：RXjava 线程调度
 */
object RxScheduler {
    /**
     * 统一线程处理
     *
     * @param <T> 指定的泛型类型
     * @return FlowableTransformer
    </T> */
    fun <T> Flo_io_main(timeOut: Int, unit: TimeUnit?): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream ->
            upstream.subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .debounce(timeOut.toLong(), unit)
        }
    }

    fun <T> Flo_io_io(timeOut: Int, unit: TimeUnit?): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream ->
            upstream.subscribeOn(Schedulers.io())
                    .observeOn(Schedulers.io())
                    .debounce(timeOut.toLong(), unit)
        }
    }

    /**
     * 统一线程处理
     *
     * @param <T> 指定的泛型类型
     * @return ObservableTransformer
    </T> */
    fun <T> Obs_io_main(timeOut: Int, unit: TimeUnit?): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream ->
            upstream.subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .debounce(timeOut.toLong(), unit)
        }
    }
}